packages/client/src/queries/applyQuery.ts

```diff
@@ -0,0 +1,330 @@
+  QueryDefinition,
+  QueryParameterDefinition,
+} from "@osdk/api";
+import type {
+  OsdkBase,
+  OsdkObjectPrimaryKeyType,
+  QueryParameterType,
+  QueryReturnType,
+} from "@osdk/client.api";
+import type {
+  DataValue,
+  ObjectSet as WireObjectSet,
+} from "@osdk/internal.foundry";
+import { OntologiesV2 } from "@osdk/internal.foundry";
+import { createAttachmentFromRid } from "../createAttachmentFromRid.js";
+import type { MinimalClient } from "../MinimalClientContext.js";
+import { createObjectSet } from "../objectSet/createObjectSet.js";
+import { addUserAgent } from "../util/addUserAgent.js";
+import { toDataValue } from "../util/toDataValue.js";
+import { toDataValueQueries } from "../util/toDataValueQueries.js";
+
+export async function applyQuery<
+  QD extends QueryDefinition<any, any>,
+  P extends QueryParameterType<QD["parameters"]>,
+>(
+  client: MinimalClient,
+  query: QD,
+  params?: P,
+): Promise<
+  QueryReturnType<QD["output"]>
+> {
+  const response = await OntologiesV2.QueryTypes.executeQueryV2(
+    addUserAgent(client, query),
+    client.ontologyRid,
+    query.apiName,
+    {
+      parameters: params
+        ? await remapQueryParams(
+          params as { [parameterId: string]: any },
+          client,
+          query.parameters,
+        )
+        : {},
+    },
+  );
+  const objectOutputDefs = await getRequiredDefinitions(query.output, client);
+  const remappedResponse = await remapQueryResponse(
+    client,
+    query.output,
+    response.value,
+    objectOutputDefs,
+  );
+  return remappedResponse as QueryReturnType<QD["output"]>;
+}
+
+async function remapQueryParams(
+  params: { [parameterId: string]: any },
+  client: MinimalClient,
+  paramTypes: Record<string, QueryParameterDefinition<any>>,
+): Promise<{ [parameterId: string]: any }> {
+  const parameterMap: { [parameterName: string]: unknown } = {};
+  for (const [key, value] of Object.entries(params)) {
+    parameterMap[key] = await toDataValueQueries(
+      value,
+      client,
+      paramTypes[key],
+    );
+  }
+  return parameterMap;
+}
+
+async function remapQueryResponse<
+  K extends string,
+  Q extends ObjectTypeDefinition<any>,
+  T extends QueryDataTypeDefinition<K, Q | never>,
+>(
+  client: MinimalClient,
+  responseDataType: T,
+  responseValue: DataValue,
+  definitions: Map<string, ObjectOrInterfaceDefinition>,
+): Promise<QueryReturnType<T>> {
+  // handle null responses
+  if (responseValue == null) {
+    if (responseDataType.nullable) {
+      return undefined as unknown as QueryReturnType<typeof responseDataType>;
+    } else {
+      throw new Error("Got null response when nullable was not allowed");
+    }
+  }
+
+  if (
+    responseDataType.multiplicity != null
+    && responseDataType.multiplicity !== false
+  ) {
+    const withoutMultiplicity = { ...responseDataType, multiplicity: false };
+    for (let i = 0; i < responseValue.length; i++) {
+      responseValue[i] = remapQueryResponse(
```

Comment:
Ah good catch, the tests for some of the nested types seemed to work fine without it, but addd the awaits in case. (Should also add a test for arrays I guess)

Reply:
Yea `responseValue[i]` is any, because it's typed as `DataValue` which is just `any`